hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -338,9 +531,14 @@ private void uploadBlockAsync(DataBlocks.DataBlock blockToUpload,
              */
             AppendRequestParameters reqParams = new AppendRequestParameters(
                 offset, 0, bytesLength, mode, false, leaseId, isExpectHeaderEnabled);
-            AbfsRestOperation op = getClient().append(path,
-                blockUploadData.toByteArray(), reqParams, cachedSasToken.get(),
-                contextEncryptionAdapter, new TracingContext(tracingContext));
+            AbfsRestOperation op;
+            try {
+              op = remoteWrite(blockToUpload, blockUploadData, reqParams, tracingContext);
+            } catch (InvalidIngressServiceException ex) {
+              switchHandler();
```

Comment:
Whether this is deterministic switching ? I meant, in case of exception what will be the behavior of the new handler?

Reply:
Also, please add debug log message in conditional execution cases.

Reply:
This fallback mechanism ensures compatibility across different endpoints. For example, if a file is created over DFS and later updated over the Blob endpoint, the operation would fail. Deterministic switching ensures that in case of an InvalidIngressServiceException, the system automatically selects the correct endpoint.